mozetl/taar/taar_similarity.py

```diff
@@ -0,0 +1,163 @@
+"""
+Bug 1386274 - TAAR similarity-based addon donor list
+
+This job clusters users into different groups based on their
+active addons. A representative users sample is selected from
+each cluster ("donors") and is saved to a model file along
+with a feature vector that will be used, by the TAAR library
+module, to perform recommendations.
+"""
+
+import click
+import json
+import logging
+
+from pyspark.sql import SparkSession
+from pyspark.ml.feature import HashingTF, IDF
+from pyspark.ml.clustering import BisectingKMeans
+from pyspark.ml import Pipeline
+from taar_utils import store_json_to_s3
+
+logging.basicConfig(level=logging.INFO)
+logger = logging.getLogger(__name__)
+
+
+def get_samples(spark):
+    """ Get a dataframe with a valid set of sample to base the next
+    processing on.
+    """
+    return spark.sql("SELECT * FROM longitudinal")\
+        .where("active_addons IS NOT null")\
+        .where("size(active_addons[0]) > 2")\
+        .where("size(active_addons[0]) < 100")\
+        .where("normalized_channel = 'release'")\
+        .selectExpr(
+            "client_id as client_id",
+            "active_addons[0] as active_addons",
+            "geo_city[0] as geo_city",
+            "subsession_length[0] as subsession_length",
+            "settings[0].locale as locale",
+            "os as os",
+            "places_bookmarks_count[0].sum AS bookmark_count",
+            "scalar_parent_browser_engagement_tab_open_event_count[0].value AS tab_open_count",
+            "scalar_parent_browser_engagement_total_uri_count[0].value AS total_uri",
+            "scalar_parent_browser_engagement_unique_domains_count[0].value AS unique_tlds"
+            )
+
+
+def get_addons_per_client(users_df, minimum_addons_count):
+    """ Extracts a dataframe that contains one row
+    for each client along with the list of active addon guids.
+    """
+    def is_valid_addon(addon):
+        if addon.is_system or\
+           addon.app_disabled or\
+           addon.type != "extension" or\
+           addon.user_disabled or\
+           addon.foreign_install:
+            return False
+
+        return True
+
+    # TODO: Better ways to do this filtering?
+    return users_df.rdd\
+        .map(lambda p: (p["client_id"],
+             [guid for guid, data in p["active_addons"].items() if is_valid_addon(data)]))\
+        .filter(lambda p: len(p[1]) > minimum_addons_count)\
+        .toDF()\
+        .selectExpr("_1 as client_id", "_2 as addon_ids")
+
+
+def compute_clusters(addons_df, num_clusters):
+    """ Performs user clustering by using addon ids as features.
+    """
+
+    # Build the stages of the pipeline. We need hashing to make the next
+    # steps work.
+    hashing_stage = HashingTF(inputCol="addon_ids", outputCol="hashed_features")
+    idf_stage = IDF(inputCol="hashed_features", outputCol="features", minDocFreq=1)
+    bkmeans_stage = BisectingKMeans(k=num_clusters)
```

Comment:
Not necessary now, but a future imporvement may be to have the termination condition relate to `minDivisibleClusterSize`. Since Bisecting K-means is a divisive clustering method, we may want to continue to sub-divide the data until the smallest single cluster contains a minimum number of elements.

Reply:
Yup, I've added a comment line there.